packages/engine/src/experiment/package/simple.rs

```diff
@@ -42,62 +63,69 @@ impl SimpleExperiment {
+                .map(|(sim_idx, props)| {
+                    // We sometimes use 0 as a default/null value, therefore it's not a valid
+                    // SimulationShortId
+                    (sim_idx + 1, props)
+                });
+
+        let mut sim_queue = SimQueue {
+            pending_iter: &mut queued_iter,
+            max_num_steps,
+            pkg_to_exp: &mut pkg_to_exp,
+            active: HashMap::new(),
+            finished: HashMap::new(),
+        };
+
+        log::trace!("Starting {max_sims_in_parallel} sims in parallel");
+        for _ in 0..max_sims_in_parallel {
+            sim_queue.start_sim_if_available().await?;
         }
 
-        // Use `isize` to avoid issues with decrementing zero.
-        let mut n_remaining = num_sims as isize;
         loop {
             let response = exp_pkg_update_recv.recv().await.ok_or_else(|| {
                 Error::ExperimentRecv(
                     "Experiment main loop closed when experiment package was still running".into(),
                 )
             })?;
 
-            let mut maybe_step_progress = n_sims_steps.get_mut(&response.sim_id);
-
             if response.was_error || response.stop_signal {
-                if let Some(step_progress) = &mut maybe_step_progress {
-                    step_progress.stopped = true;
-                } else {
-                    log::warn!("Stopped sim run with unknown id {}", &response.sim_id);
-                }
+                let mut sim_progress =
+                    sim_queue.active.remove(&response.sim_id).ok_or_else(|| {
+                        log::warn!("Sim run with unknown id {} stopped", &response.sim_id);
+                        Error::MissingSimulationRun(response.sim_id)
+                    })?;
 
-                if finish_run(&mut n_remaining) {
-                    break;
-                }
-            }
+                sim_progress.stopped = true;
+                sim_queue.finished.insert(response.sim_id, sim_progress);
 
-            let step_progress =
-                maybe_step_progress.ok_or(Error::MissingSimulationRun(response.sim_id))?;
+                sim_queue.start_sim_if_available().await?;
 
-            if !step_progress.stopped {
-                step_progress.n_steps += 1;
-            }
+                if sim_queue.active.is_empty() {
+                    break;
+                }
+            } else {
+                let mut sim_progress = sim_queue
+                    .active
+                    .get_mut(&response.sim_id)
+                    .ok_or(Error::MissingSimulationRun(response.sim_id))?;
 
-            let n_steps = step_progress.n_steps;
-            assert!(
-                n_steps <= max_num_steps,
-                "{} > max_num_steps {}",
-                n_steps,
-                max_num_steps
-            );
+                sim_progress.n_steps += 1;
 
-            if n_steps == max_num_steps && finish_run(&mut n_remaining) {
```

Comment:
Did you check that `stop_signal` is always true when a simulation run reaches the maximum number of steps? (Though it would make sense if it was true.)

Reply:
I did try and confirm that, and my manual testing has shown it to be true too. There may be some reasoning I have failed to consider when it comes to non-graceful finishing of sims, but I think we should hopefully consider that a bug rather than something that needs to be considered in this impl (also I suspect that will lead to a different component finishing that will lead to termination anyway).